packages/io/json-deserializer/index.js

```diff
@@ -1,110 +1,100 @@
-  return 'new CSG.Vertex(new CSG.Vector3D(' + ver._x + ',' + ver._y + ',' + ver._z + '))'
-}
+const { geometry } = require('@jscad/modeling')
 
-// convert the give CSG object to JSCAD source
-function toSourceCSG (csg) {
-  var code = '  var polygons = [];\n'
-  csg.polygons.map(function (p) {
-    code += '  poly = new CSG.Polygon([\n'
-    for (var i = 0; i < p.vertices.length; i++) {
-      code += '                         ' + toSourceCSGVertex(p.vertices[i].pos) + ',\n'
-    }
-    code += '                         ])'
-    if (p.shared && p.shared.color && p.shared.color.length) {
-      code += '.setColor(' + JSON.stringify(p.shared.color) + ');\n'
-    } else {
-      code += ';\n'
-    }
-    code += '  polygons.push(poly);\n'
-  })
-  code += '  return CSG.fromPolygons(polygons);\n'
-  return code
-};
-
-function toSourceCAGVertex (ver) {
-  return 'new CAG.Vertex(new CSG.Vector2D(' + ver.pos._x + ',' + ver.pos._y + '))'
-};
-function toSourceSide (side) {
-  return 'new CAG.Side(' + toSourceCAGVertex(side.vertex0) + ',' + toSourceCAGVertex(side.vertex1) + ')'
-};
-
-// convert the give CAG object to JSCAD source
-function toSourceCAG (cag) {
-  var code = '  var sides = [];\n'
-  cag.sides.map(function (s) {
-    code += '  sides.push(' + toSourceSide(s) + ');\n'
-  })
-  code += '  return CAG.fromSides(sides);\n'
-  return code
-}
+const { flatten, toArray } = require('@jscad/array-utils')
 
-// convert an anonymous CSG/CAG object to JSCAD source
-function toSource (obj) {
-  if (obj.type && obj.type === 'csg') {
-    var csg = CSG.fromObject(obj)
-    return toSourceCSG(csg)
-  }
-  if (obj.type && obj.type === 'cag') {
-    var cag = CAG.fromObject(obj)
-    return toSourceCAG(cag)
+const version = require('./package.json').version
+
+/**
+ * Deserialize the given JSON notation (string) into either a script or an array of geometry.
+ * @param {Object} [options] - options used during deserializing
+ * @param {String} [options.filename='json'] - filename of original JSON source
+ * @param {String} [options.output='script'] - either 'script' or 'geometry' to set desired output
+ * @param {String} [options.version='0.0.0'] - version number to add to the metadata
+ * @param {Boolean} [options.addMetadata=true] - toggle injection of metadata at the start of the script
+ * @param {String} input - JSON source data
+ * @return {[geometry]/String} either an array of objects (geometry) or a string (script)
+ */
+const deserialize = (options, input) => {
+  const defaults = {
+    filename: 'json',
+    output: 'script',
+    version,
+    addMetaData: true
   }
-  return ''
-};
+  options = Object.assign({}, defaults, options)
+
+  // convert the JSON notation into anonymous object(s)
+  let objects = JSON.parse(input)
+
+  // cleanup the objects
+  objects = flatten(toArray(objects))
+  objects = objects.filter((object) => geometry.geom3.isA(object) || geometry.geom2.isA(object) || geometry.path2.isA(object))
```

Comment:
should we not leave non specific jscad object 'as is' ? we have filtering in place in a ton of places, so the data should be able to be kept ?

Reply:
I added filtering in order to ensure that only JSCAD objects are returned.

I guess that non-JSCAD objects could be returned. Is this the point of your question?

Reply:
yes, in particular for JSON, it is important to be able get more data sometimes than just 3D/2D geometry, so we not filter out data unless we really need to imho :)

Reply:
@kaosat-dev do we do the same for JSON serializer as well? I’m also filtering there as well.

Reply:
Done